File: native-android/app/src/main/java/fr/eiter/plexiwine/InviteSessionStore.kt
package fr.eiter.plexiwine

import android.content.Context
import android.content.SharedPreferences
import android.util.Base64
import android.util.Log
import java.security.SecureRandom
import java.util.UUID

/**
 * Session invité APK — Bearer device-bound (WAN 4G/5G).
 * device_id stable généré une fois et renvoyé à chaque /api/native/join.
 */
object InviteSessionStore {
    private const val TAG = "InviteSession"
    private const val PREFS = "beer_invite_session_v1"
    private const val KEY_TOKEN = "access_token"
    private const val KEY_DEVICE = "device_id"
    private const val KEY_USER = "user"
    private const val KEY_LABEL = "label"
    private const val KEY_EXPIRES = "expires_at"
    private const val KEY_ACTIVE = "active"
    /** Base API (beer/ ou beer-alpha/) pour rester sur le bon backend après restart. */
    private const val KEY_API_BASE = "api_base"

    private fun prefs(context: Context): SharedPreferences =
        context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    /** Identifiant appareil stable (16–64 chars url-safe). */
    fun deviceId(context: Context): String {
        val p = prefs(context)
        val existing = p.getString(KEY_DEVICE, null)?.takeIf { it.length in 16..64 }
        if (existing != null) return existing
        val raw = ByteArray(24)
        SecureRandom().nextBytes(raw)
        val id = Base64.encodeToString(raw, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
            .take(32)
            .ifBlank { "d" + UUID.randomUUID().toString().replace("-", "").take(31) }
        p.edit().putString(KEY_DEVICE, id).apply()
        Log.i(TAG, "generated device_id len=${id.length}")
        return id
    }

    fun hasInviteSession(context: Context): Boolean {
        val p = prefs(context)
        return p.getBoolean(KEY_ACTIVE, false) &&
            !p.getString(KEY_TOKEN, null).isNullOrBlank()
    }

    fun accessToken(context: Context): String? =
        prefs(context).getString(KEY_TOKEN, null)?.takeIf { it.isNotBlank() }

    fun username(context: Context): String? =
        prefs(context).getString(KEY_USER, null)

    fun label(context: Context): String? =
        prefs(context).getString(KEY_LABEL, null)

    fun expiresAt(context: Context): String? =
        prefs(context).getString(KEY_EXPIRES, null)

    fun apiBase(context: Context): String? =
        prefs(context).getString(KEY_API_BASE, null)?.takeIf { it.isNotBlank() }

    fun save(
        context: Context,
        accessToken: String,
        user: String,
        label: String?,
        expiresAt: String?,
        deviceId: String,
        apiBase: String? = null
    ) {
        val ed = prefs(context).edit()
            .putBoolean(KEY_ACTIVE, true)
            .putString(KEY_TOKEN, accessToken)
            .putString(KEY_USER, user)
            .putString(KEY_LABEL, label)
            .putString(KEY_EXPIRES, expiresAt)
            .putString(KEY_DEVICE, deviceId)
        if (!apiBase.isNullOrBlank()) {
            ed.putString(KEY_API_BASE, ServerSettings.normalizeInput(apiBase))
        }
        ed.apply()
        Log.i(TAG, "invite session saved user=$user base=${apiBase ?: "?"}")
    }

    fun clear(context: Context) {
        val device = prefs(context).getString(KEY_DEVICE, null)
        prefs(context).edit().clear().apply()
        // Conserver le device_id pour rebind / réactivation
        if (!device.isNullOrBlank()) {
            prefs(context).edit().putString(KEY_DEVICE, device).apply()
        }
        Log.i(TAG, "invite session cleared (device kept)")
    }

    /** Extrait le token d'invite depuis une URL join ou un token brut. */
    fun parseInviteToken(raw: String): String? {
        val s = raw.trim()
        if (s.isEmpty()) return null
        // https://eiter.freeboxos.fr/wine/join/TOKEN
        val joinIdx = s.indexOf("/join/")
        if (joinIdx >= 0) {
            val after = s.substring(joinIdx + "/join/".length)
            val token = after.substringBefore('?').substringBefore('#').substringBefore('/').trim()
            return token.takeIf { it.length in 24..64 && it.matches(Regex("^[A-Za-z0-9_-]+$")) }
        }
        return s.takeIf { it.length in 24..64 && it.matches(Regex("^[A-Za-z0-9_-]+$")) }
    }
}

File: native-ios/WineNative/Sources/InviteSessionStore.swift
import Foundation
import Security

/// Session invité iOS — Bearer device-bound (WAN 4G/5G), miroir Android InviteSessionStore.
enum InviteSessionStore {
    private static let service = "fr.eiter.plexiwine.invite"
    private static let keyToken = "access_token"
    private static let keyDevice = "device_id"
    private static let keyUser = "user"
    private static let keyLabel = "label"
    private static let keyExpires = "expires_at"
    private static let keyActive = "active"
    /// Base API (beer/ ou beer-alpha/) pour rester sur le bon backend après restart.
    private static let keyApiBase = "api_base"

    private static let ud = UserDefaults.standard

    static var hasInviteSession: Bool {
        // Exige le flag UserDefaults (effacé à la désinstall) + Bearer Keychain
        // → un token Keychain orphelin après réinstall ne reconnecte pas tout seul
        guard ud.bool(forKey: keyActive) else { return false }
        return !(accessToken ?? "").isEmpty
    }

    static var accessToken: String? {
        get { keychainGet(keyToken) }
        set {
            keychainSet(keyToken, newValue)
            ud.set(newValue != nil && !(newValue ?? "").isEmpty, forKey: keyActive)
        }
    }

    static var deviceId: String {
        if let existing = keychainGet(keyDevice), existing.count >= 16, existing.count <= 64 {
            return existing
        }
        var bytes = [UInt8](repeating: 0, count: 24)
        _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        let id = Data(bytes)
            .base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
        let clipped = String(id.prefix(32))
        keychainSet(keyDevice, clipped)
        return clipped
    }

    static var username: String? {
        get { ud.string(forKey: keyUser) }
        set { ud.set(newValue, forKey: keyUser) }
    }

    static var label: String? {
        get { ud.string(forKey: keyLabel) }
        set { ud.set(newValue, forKey: keyLabel) }
    }

    static var expiresAt: String? {
        get { ud.string(forKey: keyExpires) }
        set { ud.set(newValue, forKey: keyExpires) }
    }

    static var apiBase: String? {
        get {
            guard let raw = ud.string(forKey: keyApiBase)?.trimmingCharacters(in: .whitespacesAndNewlines),
                  !raw.isEmpty else { return nil }
            return raw
        }
        set {
            if let v = newValue?.trimmingCharacters(in: .whitespacesAndNewlines), !v.isEmpty {
                ud.set(ServerSettings.normalizeInput(v), forKey: keyApiBase)
            } else {
                ud.removeObject(forKey: keyApiBase)
            }
        }
    }

    static func save(
        accessToken: String,
        user: String,
        label: String?,
        expiresAt: String?,
        deviceId: String,
        apiBase: String? = nil
    ) {
        self.accessToken = accessToken
        keychainSet(keyDevice, deviceId)
        username = user
        self.label = label
        self.expiresAt = expiresAt
        if let apiBase { self.apiBase = apiBase }
        ud.set(true, forKey: keyActive)
    }

    /// Efface le Bearer mais conserve device_id (rebind / réactivation).
    static func clear() {
        let kept = keychainGet(keyDevice)
        keychainSet(keyToken, nil)
        username = nil
        label = nil
        expiresAt = nil
        apiBase = nil
        ud.set(false, forKey: keyActive)
        if let kept { keychainSet(keyDevice, kept) }
    }

    /// Purge totale Keychain + prefs (après désinstall iOS le Keychain peut survivre).
    static func wipeAllIncludingDevice() {
        keychainSet(keyToken, nil)
        keychainSet(keyDevice, nil)
        username = nil
        label = nil
        expiresAt = nil
        apiBase = nil
        ud.set(false, forKey: keyActive)
        ud.removeObject(forKey: keyUser)
        ud.removeObject(forKey: keyLabel)
        ud.removeObject(forKey: keyExpires)
        ud.removeObject(forKey: keyActive)
    }

    /// Extrait le token depuis une URL join ou un token brut.
    static func parseInviteToken(_ raw: String) -> String? {
        let s = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !s.isEmpty else { return nil }
        if let range = s.range(of: "/join/") {
            var after = String(s[range.upperBound...])
            if let q = after.firstIndex(of: "?") { after = String(after[..<q]) }
            if let h = after.firstIndex(of: "#") { after = String(after[..<h]) }
            if let slash = after.firstIndex(of: "/") { after = String(after[..<slash]) }
            let token = after.trimmingCharacters(in: .whitespacesAndNewlines)
            return isValidToken(token) ? token : nil
        }
        return isValidToken(s) ? s : nil
    }

    private static func isValidToken(_ t: String) -> Bool {
        guard t.count >= 24, t.count <= 64 else { return false }
        return t.unicodeScalars.allSatisfy { CharacterSet.alphanumerics.contains($0) || $0 == "-" || $0 == "_" }
    }

    private static func keychainGet(_ account: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
        ]
        var item: CFTypeRef?
        guard SecItemCopyMatching(query as CFDictionary, &item) == errSecSuccess,
              let data = item as? Data,
              let value = String(data: data, encoding: .utf8),
              !value.isEmpty else { return nil }
        return value
    }

    private static func keychainSet(_ account: String, _ value: String?) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(query as CFDictionary)
        guard let value, let data = value.data(using: .utf8) else { return }
        let add: [String: Any] = query.merging([
            kSecValueData as String: data,
            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly,
        ]) { $1 }
        SecItemAdd(add as CFDictionary, nil)
    }
}
